File: tools-java-gen/src/main/kotlin/com/bytedance/ee/bear/javagen/AstParser.kt
package com.bytedance.ee.bear.javagen

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.content
import java.io.File


class AstParser(val astPath: String) {
    fun parseAst(): AstResult {
        val traitList = mutableListOf<TraitDesc>()
        val structList = mutableListOf<StructDesc>()
        val dir = File(this.astPath)
        val files = dir.listFiles();
        for (file in files) {
            val content = loadAstJson(file.absolutePath)

            val jsonElement = Json.plain.parseJson(content)
            val type = jsonElement.jsonObject["ty"].primitive.content
            if (type == "trait") {
                traitList.add(parseTrait(jsonElement))
            } else if (type == "struct") {
                structList.add(parseStruct(jsonElement))
            }
        }

        val astResult = AstResult(traitList.toTypedArray(), structList.toTypedArray())

        println("parsed json ast => $astResult")
        return astResult
    }

    private fun loadAstJson(filePath: String): String {
        println("begin load ast json for ${filePath}")
        val file = File(filePath);
        return file.readText();
    }

    private fun parseStruct(jsonElement: JsonElement): StructDesc {
        val structName = jsonElement.jsonObject["name"].primitive.content
        val modName = jsonElement.jsonObject["mod_name"].primitive.content
        val fields = mutableListOf<ArgDesc>()
        val fieldsArray = jsonElement.jsonObject["fields"].jsonArray
        for (fieldEach in fieldsArray) {
            val fieldElement = fieldEach.jsonObject
            val fieldName = fieldElement["name"].content
            val fieldOriginTy = fieldElement["origin_ty"].content
            val fieldTypeJson = fieldElement["ty"];
            val (fieldType, fieldSubType) = parseType(fieldTypeJson)
            val field = ArgDesc(fieldName, fieldType, fieldSubType, fieldOriginTy)
            fields.add(field)
        }

        return StructDesc(structName, modName, fields.toTypedArray())
    }

    private fun parseTrait(jsonElement: JsonElement): TraitDesc {
        val traitName = jsonElement.jsonObject["name"].primitive.content
        val modName = jsonElement.jsonObject["mod_name"].primitive.content
        val isCallback = jsonElement.jsonObject["is_callback"].primitive.boolean

        val methodsArray = jsonElement.jsonObject["methods"].jsonArray
        val methodList = mutableListOf<MethodDesc>()
        var index = 0
        while (index < methodsArray.size) {
            val methodElement = methodsArray[index].jsonObject
            val methodName = methodElement["name"].primitive.content
            val methodRetType = methodElement["return_type"]

            val (returnType, returnSubType) = parseType(methodRetType)
            val originRetType = methodElement["origin_return_ty"].primitive.content;

            val argArray = methodElement["args"].jsonArray

            val argList = mutableListOf<ArgDesc>()
            var i = 0
            while (i < argArray.size) {
                val argElement = argArray[i].jsonObject
                val argName = argElement["name"].content
                val argOriginTy = argElement["origin_ty"].content
                val argItem = argElement["ty"]
                val (argType, argSubType) = parseType(argItem)
                val argDesc = ArgDesc(argName, argType, argSubType, argOriginTy)
                argList.add(argDesc)
                i++
            }

            val methodDesc = MethodDesc(methodName, returnType, returnSubType, originRetType, argList.toTypedArray())
            methodList.add(methodDesc)
            index++
        }

        return TraitDesc(traitName, modName, isCallback, methodList.toTypedArray())
    }

    private fun parseType(tyElement: JsonElement): Pair<AstType, AstType> {
        val argType: AstType
        var argSubType = AstType.VOID
        if (tyElement is JsonObject) {
            val ty = tyElement.jsonObject["Vec"].content
            argType = mapType("Vec")
            argSubType = mapType(ty)
        } else {
            val ty = tyElement.content
            argType = mapType(ty)
        }

        return Pair(argType, argSubType)
    }
}

File: tools-swift-gen/Sources/SwiftGenCore/AstParser.swift
//
//  AstParser.swift
//  SwiftGenCore
//
//  Created by wangxin.sidney on 2018/6/22.
//

import Foundation

class AstParser {
    let astPath: String
    
    init(path: String) {
        self.astPath = path
    }
    
    public func parseAst() -> AstResult {
        let dic = loadAstJson()
        var traits: [TraitDesc] = []
        var structs: [StructDesc] = []
        
        let ty = dic["ty"] as! String
        if ty == "struct" {
            let struct_ = parseStruct(dic: dic)
            structs.append(struct_)
        } else if ty == "trait" {
            let trait = parseTrait(dic: dic)
            traits.append(trait)
        }
        
        return AstResult(traits: traits, structs: structs)
    }
    
    private func parseStruct(dic: [String:Any]) -> StructDesc {
        let structName = dic["name"] as! String
        let modName = dic["mod_name"] as! String
        let crateName = dic["crate_name"] as! String
        let fields = dic["fields"] as! [Any]
        let args = parseArgs(args: fields)
        
        return StructDesc(name: structName, mod_name: modName, crate_name:crateName,  fields: args)
    }
    
    private func parseTrait(dic: [String:Any]) -> TraitDesc {
        let traitName = dic["name"] as! String
        let modName = dic["mod_name"] as! String
        let crateName = dic["crate_name"] as! String
        let methods = dic["methods"] as! [Any]
        let isCallback = dic["is_callback"] as! Bool
        
        var methodDescs: [MethodDesc] = []
        for method in methods {
            let method = method as! [String: Any]
            
            let methodName = method["name"] as! String
            let orginReturnType = method["origin_return_ty"] as! String
            let args = method["args"] as! [Any]
            
            var returnType = AstType.VOID
            if let returnTy = method["return_type"] as? String {
                returnType = AstType.fromStr(ty: returnTy, originTy: orginReturnType)
            } else {
                if let vecType = method["return_type"] as? [String: String] {
                    let baseType = vecType["Vec"]!
                    returnType = AstType.VEC(AstBaseType.fromStr(str: baseType))
                }
            }
            
            var argDescs:[ArgDesc] = parseArgs(args: args)
            let methodDesc = MethodDesc(name: methodName, return_type: returnType, origin_return_ty: orginReturnType, args: argDescs)
            methodDescs.append(methodDesc)
        }
        let traitDesc = TraitDesc(name: traitName, mod_name: modName, crate_name:crateName, is_callback: isCallback,  methods: methodDescs)
        return traitDesc
    }
    
    private func parseArgs(args: [Any]) -> [ArgDesc] {
        var argDescs:[ArgDesc] = []
        for arg in args {
            let arg = arg as! [String: Any]
            let argName = arg["name"] as! String
            let (argType, argOriginType) = parseType(arg: arg)
            let argDesc = ArgDesc(name: argName, ty: argType, origin_ty: argOriginType)
            argDescs.append(argDesc)
        }
        
        return argDescs
    }
    
    private func parseType(arg: [String: Any]) -> (AstType, String) {
        let argOriginType = arg["origin_ty"] as! String
        var argType = AstType.VOID
        if let argTy = arg["ty"] as? String  {
            argType = AstType.fromStr(ty: argTy, originTy: argOriginType)
        } else {
            if let vecType = arg["ty"] as? [String: String] {
                let baseType = vecType["Vec"]!
                argType = AstType.VEC(AstBaseType.fromStr(str: baseType))
            }
        }
        
        return (argType, argOriginType)
    }
    
    private func loadAstJson() -> [String:Any] {
        print("begin load ast json for \(self.astPath).")
        
        let fileManager = FileManager.default
        if let jsonData = fileManager.contents(atPath: self.astPath) {
            do {
                if let dic = try JSONSerialization.jsonObject(with: jsonData, options: JSONSerialization.ReadingOptions(rawValue: 0)) as? [String: Any] {
                    return dic
                } else {
                    print("parse ast json to empty!")
                    return [:]
                }
            } catch {
                print("parse ast json wrong!")
            }
        } else {
            print("obtain data from file \(self.astPath) error.")
        }
        
        return [:]
    }
}
